[VFABI] Create FunctionType for vector functions

paschalis-mpeis opened the pull request:
`createFunctionType` optionally returns a FunctionType and the mask's position when there's one. It requires VFInfo and an Instruction.

Add `checkFunctionType` in 'VectorFunctionABITest.cpp' tests to check that both the number and the type of vectorized parameters matches the created `FunctionType`.

paschalis-mpeis committed:
[VFABI] Create FunctionType for vector functions

`createFunctionType` optionally returns a FunctionType and the mask's
position when there's one. It requires VFInfo and an Instruction.

Add `checkFunctionType` in 'VectorFunctionABITest.cpp' tests to check
that both the number and the type of vectorized parameters matches the
created `FunctionType`.

paschalis-mpeis committed:
Handle already masked Instructions.

'createFunctionType' should be able to create the correct FunctionType,
regardless of whether the input Instruction was masked or not.
It uses VFInfo to figure out if the input Instruction was already
masked, and if so it does not append another mask Type.

In checks, create two mock CallInsts, one with a mask and one without
and verify that they have the same number of parameters.

paschalis-mpeis committed:
Simplified createFunctionType method

It accepts ScalarFTy and VecRetTy. As the latter is not kept in VFABI,
it should instead come from the original Instruction/Callinst.

This change allows further simplification in tests. Also, methods like
the test `_ZGVnN3lLRUlnLnRnUn_foo`, which contains only scalar parameters,
would still return a valid FunctionType. This is an edge case that will
only be encountered in tests though.

paulwalker-arm commented on llvm/include/llvm/Analysis/VectorUtils.h:
```diff
@@ -195,6 +195,12 @@ static constexpr char const *MappingsAttrName = "vector-function-abi-variant";
+
+/// Returns a pair of the vectorized FunctionType and the mask's position when
+/// there's one, otherwise -1.
+std::optional<std::pair<FunctionType *, int>>
```
What's the rational for conflating these two properties given `VFInfo` already has a nice interface to query for a mask's presences and position.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1479,34 @@ void VFABI::getVectorVariantNames(
 
+std::optional<std::pair<FunctionType *, int>>
+VFABI::createFunctionType(const VFInfo &Info, const FunctionType *ScalarFTy,
+                          Type *VecRetTy, const Module *M) {
```
Shouldn't the return type of the vector function also be constructed from `ScalarFTy`?

paschalis-mpeis replied:
Discussed in person and decided to drop `VecRetTy`.

`VFInfo` does not hold such information since mangled names do not encode it,
however, if that changes then it will. So one should be able to get it from there.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1479,34 @@ void VFABI::getVectorVariantNames(
+      return std::nullopt;
+
+    MaskPos = OptMaskPos.value();
+    VectorType *MaskTy = VectorType::get(Type::getInt1Ty(M->getContext()), VF);
```
You should be able to extract the context from `ScalarFTy` rather than need to pass in the module.

paschalis-mpeis committed:
createFunctionType requires only VFInfo and ScalarFTy

Not returning a pair anymore as the position can be queries directly from
VFInfo, which createFunctionType needs to have as an argument to begin
with.

Also getting the return type from the ScalarFTy, as the specification
does not encode in the mangled name such information. Therefore, the
VFInfo does not hold such info. If that changes, then it will make its
way into VFInfo and one could get it from there.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,37 @@ void VFABI::getVectorVariantNames(
   }
 }
 
+std::optional<FunctionType *>
```
Can this just be `FunctionType *`? We don't need to distinguish between `std::nullopt` and `nullptr` so I don't think the optional provides any value.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,37 @@ void VFABI::getVectorVariantNames(
+
+  // Get mask's position mask and append one if not present in the Instruction.
+  int MaskPos = -1;
+  if (Info.isMasked()) {
+    auto OptMaskPos = Info.getParamIndexForOptionalMask();
```
What about `if (auto OptMaskPos = Info.getParamIndexForOptionalMask())`?

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,37 @@ void VFABI::getVectorVariantNames(
+  }
+
+  // Get mask's position mask and append one if not present in the Instruction.
+  int MaskPos = -1;
```
I don't think this is needed anymore?

paschalis-mpeis committed:
Cleanup: addressing review.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,32 @@ void VFABI::getVectorVariantNames(
+
+  // Get mask's position mask and append one if not present in the Instruction.
+  if (auto OptMaskPos = Info.getParamIndexForOptionalMask()) {
+    if (!OptMaskPos)
+      return nullptr;
```
This code is unreachable and can be removed.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,32 @@ void VFABI::getVectorVariantNames(
+      VecTypes.push_back(STy);
+  }
+
+  // Get mask's position mask and append one if not present in the Instruction.
```
I'm struggling to parse the comment.  How about "Insert mask operand if required."?

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,32 @@ void VFABI::getVectorVariantNames(
+  }
+  auto *RetTy = ScalarFTy->getReturnType();
+  if (!RetTy->isVoidTy())
+    RetTy = VectorType::get(ScalarFTy->getReturnType(), VF);
```
Could reuse `RetTy` here instead of calling `getReturnType()` again.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -20,10 +21,12 @@
 #include "llvm/IR/PatternMatch.h"
 #include "llvm/IR/Value.h"
 #include "llvm/Support/CommandLine.h"
+#include <optional>
```
You don't directly use this so can rely on it being brought in via whatever defines `VFInfo`.

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1480,32 @@ void VFABI::getVectorVariantNames(
+  ElementCount VF = Info.Shape.VF;
+  // Create vector parameter types
+  SmallVector<Type *, 8> VecTypes;
+  for (auto [STy, VFParam] : zip(ScalarFTy->params(), Info.Shape.Parameters)) {
```
This only works because we know any mask parameter will be last?

Given the following code `VecTypes.insert(VecTypes.begin() + OptMaskPos.value(), MaskTy);` allows the mask to be anywhere I think we'd be better served being more explicit. For example:

```
ScalarParamIndex = 0;
foreach (Info.Shape.Parameters) {
  if (ParamKind == GlobalPredicate) {
    VecTypes.push_back(mask_type)
    continue;
  }

  OperandType = getParamType(ScalarParamIndex++)
  if (ParamKind == vector) {
    OperandType = vector_type(OperandType, VF)

  VecTypes.push_back(OperandType)
}
```
Perhaps there's a nicer c++ way to write this but you get the gist?

paschalis-mpeis replied:
> This only works because we know any mask parameter will be last?

True, because specs guarantees that.
Thanks for the gist, looks better and it'll handle things nicely.

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -91,6 +92,75 @@ class VFABIParserTest : public ::testing::Test {
+
+    FunctionCallee F = M->getOrInsertFunction(
+        VectorName, FunctionType::get(VecRetTy, CallTypes, false));
+    std::unique_ptr<CallInst> CI(CallInst::Create(F, Args));
```
Is CI used anywhere?

Regardless of this, it doesn't make sense to me to verify the quality of `createFunctionType` by implementing the same algorithm again or by comparing to the input data used to create the function type.  I would expect the expected result to be pass in explicitly and then compared.

Is it possible to have `EXPECT_EQ(createFunctionType(..), TwoOpMaskedFuncTy)` with a bunch of global FunctionTypes?  I'm assuming there's a small subset of FunctionTypes you need, otherwise I guess they should be declared locally.

paschalis-mpeis committed:
Improve testing for createFunctionType.

Refactored 'createFunctionType'

paulwalker-arm commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -1477,6 +1479,31 @@ void VFABI::getVectorVariantNames(
+      continue;
+    }
+
+    Type *OperandTy = ScalarFTy->getParamType(Idx);
```
As before I think this is only safe because the global predicate is last.  This is why my example maintained a separate parameter count for the scalar function parameters.

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -318,7 +392,7 @@ TEST_F(VFABIParserTest, Align) {
   EXPECT_EQ(ISA, VFISAKind::SVE);
   EXPECT_FALSE(isMasked());
-  EXPECT_TRUE(matchParametersNum()) << "Different number of scalar parameters";
+  EXPECT_TRUE(matchParametersNum());
```
Does this need a matching `EXPECT_EQ(getFunctionType,...)`?

paschalis-mpeis committed:
Improve handling of Vectors in createFunctionType.

If the VFABI specification changes, allowing vectors in any position
other than the end of the argument list, createFunctionType should now
be able to handle it.

mgabka commented on llvm/include/llvm/Analysis/VectorUtils.h:
```diff
@@ -195,6 +195,11 @@ static constexpr char const *MappingsAttrName = "vector-function-abi-variant";
+
+/// Constructs a FunctionType by applying vector function information to the
+/// type of a matching scalar function.
+FunctionType *createFunctionType(const VFInfo &Info,
```
nit: since this is going to be publicly available interface, I think it is worth to use doxygen style to document the parmeters.

paschalis-mpeis committed:
Updated docs.
